overlord/configstate/config/transaction.go

```diff
@@ -361,3 +485,35 @@ func (e *NoOptionError) Error() string {
 	}
 	return fmt.Sprintf("snap %q has no %q configuration option", e.SnapName, e.Key)
 }
+
+// VirtualCfgFunc can be used for virtual "transaction.Get()" calls
+type VirtualCfgFunc func(snapName, key string) (result interface{}, err error)
+
+// virtualMap contain hook functions for "virtual" configuration. The
+// first level of the map is the snapName and then the virtual keys in
+// doted notation e.g. "network.netplan".  Any data under a virtual
+// configuration option is never stored directly in the state.
+var (
+	virtualMap map[string]map[string]VirtualCfgFunc
+	virtualMu  sync.Mutex
+)
+
+// RegisterVirtualConfig allows to register a function that is called
+// when the configuration for the given config key for a given
+// snapname is requested.
+//
+// This is useful for e.g. the system.hostname configuration where the
+// authoritative value is coming from the kernel and can be change
+// outside of snapd.
+func RegisterVirtualConfig(snapName, key string, hi VirtualCfgFunc) {
+	virtualMu.Lock()
+	defer virtualMu.Unlock()
+
+	if virtualMap == nil {
+		virtualMap = make(map[string]map[string]VirtualCfgFunc)
+	}
+	if _, ok := virtualMap[snapName]; !ok {
+		virtualMap[snapName] = make(map[string]VirtualCfgFunc)
+	}
+	virtualMap[snapName][key] = hi
```

Comment:
any reason why this is called "hi"?

Reply:
Good point, renamed to `vf` (virtual function) :)